app/attributes/attributecontroller.cpp

```diff
@@ -875,14 +919,26 @@ bool AttributeController::setFormValue( const QUuid &id, QVariant value )
         QString msg( tr( "Value \"%1\" %4 could not be converted to a compatible value for field %2(%3)." ).arg( value.toString(), fld.name(), fld.typeName(), value.isNull() ? "NULL" : "NOT NULL" ) );
         QString userFriendlyMsg( tr( "Value %1 is not compatible with field type %2." ).arg( value.toString(), fld.typeName() ) );
         QgsMessageLog::logMessage( msg );
-        if ( !val.isNull() )
+
+        if ( value.toBool() )
         {
-          emit formDataChangedFailed( userFriendlyMsg );
+          item->setState( FormItem::InvalidInput );
+          emit formDataChanged( id, { AttributeFormModel::FieldState } );
+
+//          I vote for removing this toast message. It did not work with recent versions and nobody seems to care
+//          so I would just remove it
+//          It was added for QR code, we can add similar message like this min/max range instead
+//          emit formDataChangedFailed( userFriendlyMsg );
```

Comment:
This is the toast notification. It was not working in previous versions

Reply:
+1 better to use text below field similar to constrains